core/src/main/java/org/infinispan/scattered/impl/ScatteredStateConsumerImpl.java

```diff
@@ -0,0 +1,527 @@
+         if (!newWriteCh.isSegmentLocalToNode(rpcManager.getAddress(), segment)) {
+            svm.unregisterSegment(segment);
+         }
+      }
+      Set<Integer> addedSegments = getOwnedSegments(newWriteCh);
+      if (previousWriteCh != null) {
+         addedSegments.removeAll(getOwnedSegments(previousWriteCh));
+      }
+      svm.setTopologyId(topologyId);
+      if (previousWriteCh == null || !isFetchEnabled) {
+         log.trace("This is the first topology or state transfer is disabled, not expecting any state transfer.");
+         svm.setOwnedSegments(addedSegments);
+         return;
+      }
+      if (!addedSegments.isEmpty()) {
+         svm.setValuesTransferTopology(topologyId);
+      }
+      for (int segment : addedSegments) {
+         svm.registerSegment(segment);
+      }
+   }
+
+   @Override
+   protected void handleSegments(boolean startRebalance, Set<Integer> addedSegments, Set<Integer> removedSegments) {
+      if (!startRebalance) {
+         log.trace("This is not a rebalance, not doing anything...");
+         return;
+      }
+      if (addedSegments.isEmpty()) {
+         log.trace("No segments missing");
+         return;
+      }
+
+      synchronized (transferMapsLock) {
+         inboundSegments = new HashSet<>(addedSegments);
+      }
+      chunkCounter.set(0);
+
+      rpcManager.invokeRemotelyAsync(null, commandsFactory.buildStateRequestCommand(
+            StateRequestCommand.Type.CONFIRM_REVOKED_SEGMENTS,
+            rpcManager.getAddress(), cacheTopology.getTopologyId(), null),
+            synchronousIgnoreLeaversRpcOptions).whenComplete((responses, throwable) -> {
+         if (throwable == null) {
+            try {
+               svm.startKeyTransfer(addedSegments);
+               requestKeyTransfer(addedSegments);
+            } catch (Throwable t) {
+               log.failedToRequestSegments(cacheName, null, addedSegments, t);
+            }
+         } else {
+            if (cache.getAdvancedCache().getComponentRegistry().getStatus() == ComponentStatus.RUNNING) {
+               log.failedConfirmingRevokedSegments(throwable);
+            } else {
+               // reduce verbosity for stopping cache
+               log.debug("Failed confirming revoked segments", throwable);
+            }
+            for (int segment : addedSegments) {
+               svm.notifyKeyTransferFinished(segment, false);
+            }
+            notifyEndOfStateTransferIfNeeded(cacheTopology.getTopologyId(), cacheTopology.getRebalanceId());
+         }
+      }).join(); // we need to wait synchronously for the completion
+   }
+
+   private void requestKeyTransfer(Set<Integer> segments) {
+      boolean isTransferingKeys = false;
+
+      synchronized (transferMapsLock) {
+         List<Address> members = new ArrayList<>(cacheTopology.getActualMembers());
+         // Reorder the member set to distibute load more evenly
+         Collections.shuffle(members);
```

Comment:
You have to contact all the nodes, but you don't want all A B C D contacting first A, then all contacting B...